### ISSUE TITLE
[Workspace Feeds] [$250] Card - "Account ending in" message appears after adding bank account for the card

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#45729]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
Issue found when validating #45388 
**Version Number:** 9.0.9-1
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause Internal Team
**Slack conversation:**

## Action Performed:
1. Navigate to https://staging.new.expensify.com/
2. Log in with a new expensifail account
3. Create a workspace and set default currency to USD
4. Navigate to Workspace settings - More features
5. Enable Expensify Card
6. Navigate to Expensify Card
7. Click on "Issue new card" button
8. Click on "Connect online with Plaid"
9. Complete BA flow with Regions bank
10. Click on "Issue new card" button
## Expected Result:
The account holders name and the accounts last four number should be showing.
## Actual Result:
"Account ending in" message appears after adding bank account for the card. You have to relog for the correct info to load.

## Workaround:
Unknown
## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

https://github.com/user-attachments/assets/6ad8a5a7-6e4c-48ba-91ff-c29784f21626

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01622f31aad5cbea6e</li>
        <li>Upwork Job ID: 1814111023366877998</li>
        <li>Last Price Increase: 2024-07-26</li>
    </ul>
</details>

**Comment by m-natarajan** [comment]
@kadiealexander FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by m-natarajan** [comment]
We think that this bug might be related to #wave-collect - Release 2

**Comment by VasylMartyniv** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue.
Account details not being correctly displayed after adding new account. The account's data is not being passed properly to the item in a list.

### What is the root cause of that problem?
The eligibleBankAccounts mapping function that maps bankAccountList to menu items, is trying to use some values from `bankAccount.accountData` which are not present in that object after adding the account.

### What changes do you think we should make in order to solve the problem?
For displaying the data in menu item we can use the values from `bankAccountList` item iteself. So the proposed changes are next: 
- For the display of Name and Account number use the data from `bankAccount` object, and not from `bankAccount.accountData`.
    - Use `bankAccount.description.slice(-4)` for the `bankAccountNumber` variable
    - Use `bankAccount.title` for the Account Holder name in the menu item
- Because the `bankName` property is present directly in the `accountData` object after adding a new account, we can add a conditional that will use the `bankName` variable from `accountData.bankName`, and if that is missing then use a value from `accountData.additionalData.bankName`.
- Adjust "AccountData" type, as the property `bankName` can be present both in `AccountData` type or the `BankAccountAdditionalData` type, depending on the response that provides `bankAccountList`.

All of the above variables are already accessible in `bankAccountList` items after adding a new account, so no additional fetching needed.

### What alternative solutions did you explore? (Optional)
- Additional fetching of the missing data after adding new card
- Adjusting the back-end responses to have a consistent bankAccount object structure that they return.

### Result of proposed changes:
<img width="375" alt="Screenshot 2024-07-22 at 4 31 15 PM" src="https://github.com/user-attachments/assets/02445bd7-367f-4223-b090-3b23e17bce71">

**Comment by VasylMartyniv** [comment]
Contributor details
Your Expensify account email: vasyl.martyniv@perfsol.tech
Upwork Profile Link: https://www.upwork.com/freelancers/~018d00e8b30180eee9

**Comment by s77rt** [comment]
@nyomanjyotisa

> the `bankAccounts` onyx data given by the API not formatted properly

Can you elaborate how not formatted? (include returned data vs expected data)

**Comment by s77rt** [comment]
@VasylMartyniv 

> trying to use some values from `bankAccount.accountData` which are not present in that object after adding the account

Why do we have missing fields? And are the rest of the fields returned by the backend or populated optimistically?

**Comment by s77rt** [comment]
Waiting on proposals

**Comment by koko57** [comment]
@s77rt I am still working on Expensify Card (Workspace Feed) project and I was working on the adding bank account logic. Maybe I could work on that?

**Comment by s77rt** [comment]
🎀 👀 🎀 Please assign @koko57

**Comment by iwiznia** [comment]
Shouldn't there be a proposal?

**Comment by s77rt** [comment]
@iwiznia agency-employees are assigned immediately then a proposal is submitted

**Comment by s77rt** [comment]
Not overdue. Waiting on proposal

**Comment by koko57** [comment]
I will be working on this today

**Comment by koko57** [comment]
I'm trying to reproduce the bug, but even with providing the data to skip the Verification wit Onfido, or trying to verify I can't get my account added to the bankAccount list. But I agree with previous proposals - it would be good to check why the bankAccount is not returned after adding it the same way the same bank account is then returned with `openWalletPage`

**Comment by s77rt** [comment]
No update yet. Still having issues reproducing this due to verification flow
- https://expensify.slack.com/archives/C01GTK53T8Q/p1721668991138459?thread_ts=1711049879.962529&cid=C01GTK53T8Q
- https://expensify.slack.com/archives/C01GTK53T8Q/p1723019183194709?thread_ts=1711049879.962529&cid=C01GTK53T8Q

**Comment by koko57** [comment]
I will investingate it today

**Comment by koko57** [comment]
posted my findings here: https://github.com/Expensify/App/issues/47239#issuecomment-2286312613

**Comment by s77rt** [comment]
Same ^

**Comment by s77rt** [comment]
Same ^

**Comment by koko57** [comment]
@s77rt more explaination here: https://github.com/Expensify/App/issues/47239#issuecomment-2298836350

The issue with the data not being sent the same way is solved, but we still need to make one fix here.
We need to change `bankAccount.accountData?.accountNumber` for `bankAccount.accountData?.additionalData?.mask` 

https://github.com/Expensify/App/blob/a32da40e1017d2136389f05735bace7bc307e098/src/pages/workspace/expensifyCard/WorkspaceExpensifyCardBankAccounts.tsx#L51

and use the mask without additionally using `getLastFourDigits`

https://github.com/Expensify/App/blob/a32da40e1017d2136389f05735bace7bc307e098/src/pages/workspace/expensifyCard/WorkspaceExpensifyCardBankAccounts.tsx#L59

**Comment by s77rt** [comment]
@koko57 Thanks! Let's wait for @mountiny on this one, for consistency I think we may want the `accountNumber`

**Comment by koko57** [comment]
@mountiny waiting for your decision here 🙂 Personally I would go with changing this to `mask` regardless if the `accountNumber` is added or not, because we could get rid of `getLastFourDigits` calculation and `mask` is something that should be returned for every card

**Comment by mountiny** [comment]
We should work on unifying how we send the data from those two commands, I created a task to do that here https://github.com/Expensify/App/issues/47944


I got a quick PR to add the masked `accountNumber` to the response here too so we can keep the behaviour unified.

**Comment by koko57** [comment]
@mountiny should I make this change with mask that I proposed anyway? or are we going with just the backend fix?

**Comment by mountiny** [comment]
Lets just wait for the backend fix so we do not handle this differently in two places

**Comment by mountiny** [comment]
@koko57 the fix is deployed, can you retest it?

**Comment by koko57** [comment]
the same as in https://github.com/Expensify/App/issues/47239#issuecomment-2312597635

**Comment by kadiealexander** [comment]
Not overdue

**Comment by mountiny** [comment]
Follow up backend PR merged

**Comment by kadiealexander** [comment]
@mountiny is there anything else left to complete on this issue?

**Comment by mountiny** [comment]
@koko57 We had to make one small backend change and that should have been deployed. Can you please confirm if there is anything left to do for this one? Thank you 🙇

**Comment by mountiny** [comment]
Perfect, this is fixed now then, thank you for confirming!
